### DISCUSSION TITLE
Customized onchange event for controlled input in material-ui

### DISCUSSION THREAD

**Discussion by mustanish** [react-hook-form/react-hook-form#3387] [Q&A] [answered]
I have used react material-ui to create a form. if I do not use RHF then I am able to change the state of a component, but how I achieve the same with RHF. 

For example, if my textField component is something like this
`<TextField name="firstName" label="First Name" type="text" required={true}  value={user.firstName} onChange={onChange}/>`
Then how do I write equivalent with RHF

**Comment by bluebill1049** [comment, marked as answer]
1 `onChange` with `setValue`
2. Controller https://codesandbox.io/s/react-hook-form-v6-validationschema-forked-jopm9?file=/src/index.js

**Reply by mustanish** [reply]
The logic of the form is somewhat like that it comes prefilled if we have data else it is rendered empty

```
const profile = useSelector((state: RootState) => state.profile);

const [user, setUser] = React.useState<ProfileType>(profile);
  
const { handleSubmit, errors, control } = useForm<ProfileType>({defaultValues:{...user}, resolver: joiResolver(ProfileSchema)});

  const onSubmit = (data: ProfileType) => {
    dispatch(setProfile(data));
    setUser(Object.assign(user, data))
    history.push('/question');
  }

  const onChange = (event: any) => {
    console.log(event.target.name);
    console.log(event.target.value);
    setUser({ ...user, [event.target.name]: event.target.value });
  };

  React.useEffect(() => {
    setUser(profile);
  }, [profile]);
```

Under the return method I have a form using material-UI with RHF, As you can see that with useEffect form is being rerendered when the value of the profile is changed. The value comes prefilled under form fields using material-ui but for fields using the RHF(Controller), it does not come prefilled.

For example, my approach to using RHF is as below
`<Controller render={(props) => ( <TextField value={props.value} onChange={props.onChange}/>)} name="firstName" control={control}/>
 <TextField name="firstName" label="First Name" type="text"  value={user.firstName} onChange={onChange}/>`

For TextField value of firstName changes from empty (Initial value) to newValue after useEffect finishes running when the value of profile is updated from the global store. How do I achieve something with RHF? I hope I have explained my issue.
